File: daemon/src/events.rs
//! Event envelope mirroring contracts/event_stream.schema.json (tos.app.event.v0).
//! The JSON schema is the law; this struct is its Rust projection. Contract
//! conformance is enforced by tests that replay every committed fixture.

use serde::{Deserialize, Serialize};

pub const EVENT_SCHEMA_VERSION: &str = "tos.app.event.v0";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EventEnvelope {
    pub event_id: String,
    pub seq: u64,
    pub ts: String,
    pub schema_version: String,
    pub kind: EventKind,
    pub source: EventSource,
    pub trust_state: TrustState,
    pub payload: serde_json::Value,
}

impl EventEnvelope {
    /// The one envelope constructor (A1_03: a second hand-rolled copy is how
    /// paired paths drift). event_id derives from the sanitized project id +
    /// seq and always matches the schema pattern `^evt_[a-z0-9_]+$`.
    pub fn new(
        project_id: &str,
        seq: u64,
        ts: &str,
        kind: EventKind,
        source: EventSource,
        trust_state: TrustState,
        payload: serde_json::Value,
    ) -> Self {
        EventEnvelope {
            event_id: format!("evt_{}_{seq:04}", sanitize_id(project_id)),
            seq,
            ts: ts.to_string(),
            schema_version: EVENT_SCHEMA_VERSION.to_string(),
            kind,
            source,
            trust_state,
            payload,
        }
    }
}

/// Lowercase [a-z0-9_] identifier component (schema id patterns).
pub(crate) fn sanitize_id(s: &str) -> String {
    let mut out: String = s
        .chars()
        .map(|c| {
            let c = c.to_ascii_lowercase();
            if c.is_ascii_lowercase() || c.is_ascii_digit() {
                c
            } else {
                '_'
            }
        })
        .collect();
    if out.is_empty() {
        out.push('x');
    }
    out
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EventKind {
    ProjectRegistered,
    WorktreeDiscovered,
    WorktreeRemoved,
    FileChanged,
    DiffSnapshot,
    ProposalCandidate,
    ProposalSubmitted,
    ProposalRejected,
    ProposalAccepted,
    PredicateResult,
    VetoVerdict,
    AgentManifestRegistered,
    AgentSessionStarted,
    AgentSessionEnded,
    SignatureVerified,
    SignatureRejected,
    RatificationProposed,
    RatificationCeremonyOpened,
    RatificationSigned,
    RatificationTagCreated,
    MarketTxObserved,
    ReconciliationCompleted,
    BranchObserved,
    BranchRemoved,
    CommitObserved,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EventSource {
    Git,
    Fsevents,
    ClaudeHook,
    CodexAppserver,
    Human,
    Daemon,
    Fixture,
    Github,
}

/// ActorTrustState - the single trust vocabulary (docs/TRUST_STATES.md).
/// Machine source of truth is contracts/event_stream.schema.json; changes
/// must touch both and pass shipgate.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TrustState {
    ObservedUnsigned,
    ManifestMissing,
    ManifestRegistered,
    SignatureValid,
    SignatureInvalid,
    SignerUnregistered,
    SignerRevoked,
    CapabilityMissing,
    HumanAdopted,
    HumanRootSigned,
    LegacyPreRule,
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;

    fn fixtures_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../fixtures/event_streams")
    }

    /// Repo law replayed in Rust: every committed fixture line must
    /// deserialize into the envelope with strictly increasing seq and the
    /// pinned schema_version. If this fails, either the contract drifted or
    /// this projection did - both are gate-red conditions.
    #[test]
    fn fixtures_conform_to_envelope() {
        let mut files: Vec<_> = fs::read_dir(fixtures_dir())
            .expect("fixtures dir")
            .map(|e| e.expect("entry").path())
            .filter(|p| p.extension().is_some_and(|x| x == "jsonl"))
            .collect();
        files.sort();
        assert!(!files.is_empty(), "no fixtures found - repo law missing");
        for file in files {
            let body = fs::read_to_string(&file).expect("read fixture");
            let mut prev_seq: Option<u64> = None;
            for (n, line) in body.lines().enumerate() {
                if line.trim().is_empty() {
                    continue;
                }
                let ev: EventEnvelope = serde_json::from_str(line)
                    .unwrap_or_else(|e| panic!("{}:{} bad envelope: {e}", file.display(), n + 1));
                assert_eq!(
                    ev.schema_version,
                    EVENT_SCHEMA_VERSION,
                    "{}",
                    file.display()
                );
                if let Some(p) = prev_seq {
                    assert!(
                        ev.seq > p,
                        "{}: seq not strictly increasing",
                        file.display()
                    );
                }
                prev_seq = Some(ev.seq);
            }
        }
    }

    #[test]
    fn trust_state_round_trips_snake_case() {
        let js = "\"human_root_signed\"";
        let t: TrustState = serde_json::from_str(js).unwrap();
        assert_eq!(t, TrustState::HumanRootSigned);
        assert_eq!(serde_json::to_string(&t).unwrap(), js);
    }
}

File: app/Sources/TuringOS/Events.swift
// Swift projection of contracts/event_stream.schema.json (tos.app.event.v0).
// The JSON schema is the law; this mirrors daemon/src/events.rs. Contract
// conformance is enforced by tests that replay every committed fixture -
// the same repo-law replay the Rust side runs, so the two language mirrors
// cannot drift silently (cross-language conservation via shared fixtures).
//
// Strictness asymmetry is DELIBERATE (S-stage finding, ratified here): the
// Rust PRODUCER mirror is deny_unknown_fields (tightness enforced where
// envelopes are born); this CONSUMER mirror tolerates additive unknown
// fields so the app survives daemon minor-version field additions without
// a lockstep release. Value/count drift is still caught by the shared
// fixtures; only consumer-side looseness is intentionally open.

import Foundation

public let eventSchemaVersion = "tos.app.event.v0"

public struct EventEnvelope: Codable, Sendable, Equatable {
    public let eventId: String
    public let seq: UInt64
    public let ts: String
    public let schemaVersion: String
    public let kind: EventKind
    public let source: EventSource
    public let trustState: TrustState
    public let payload: JSONValue

    enum CodingKeys: String, CodingKey {
        case eventId = "event_id"
        case seq, ts
        case schemaVersion = "schema_version"
        case kind, source
        case trustState = "trust_state"
        case payload
    }
}

public enum EventKind: String, Codable, Sendable, CaseIterable {
    case projectRegistered = "ProjectRegistered"
    case worktreeDiscovered = "WorktreeDiscovered"
    case worktreeRemoved = "WorktreeRemoved"
    case fileChanged = "FileChanged"
    case diffSnapshot = "DiffSnapshot"
    case proposalCandidate = "ProposalCandidate"
    case proposalSubmitted = "ProposalSubmitted"
    case proposalRejected = "ProposalRejected"
    case proposalAccepted = "ProposalAccepted"
    case predicateResult = "PredicateResult"
    case vetoVerdict = "VetoVerdict"
    case agentManifestRegistered = "AgentManifestRegistered"
    case agentSessionStarted = "AgentSessionStarted"
    case agentSessionEnded = "AgentSessionEnded"
    case signatureVerified = "SignatureVerified"
    case signatureRejected = "SignatureRejected"
    case ratificationProposed = "RatificationProposed"
    case ratificationCeremonyOpened = "RatificationCeremonyOpened"
    case ratificationSigned = "RatificationSigned"
    case ratificationTagCreated = "RatificationTagCreated"
    case marketTxObserved = "MarketTxObserved"
    case reconciliationCompleted = "ReconciliationCompleted"
    case branchObserved = "BranchObserved"
    case branchRemoved = "BranchRemoved"
}

public enum EventSource: String, Codable, Sendable {
    case git, fsevents, human, daemon, fixture, github
    case claudeHook = "claude_hook"
    case codexAppserver = "codex_appserver"
}

public enum TrustState: String, Codable, Sendable {
    case observedUnsigned = "observed_unsigned"
    case manifestMissing = "manifest_missing"
    case manifestRegistered = "manifest_registered"
    case signatureValid = "signature_valid"
    case signatureInvalid = "signature_invalid"
    case signerUnregistered = "signer_unregistered"
    case signerRevoked = "signer_revoked"
    case capabilityMissing = "capability_missing"
    case humanAdopted = "human_adopted"
    case humanRootSigned = "human_root_signed"
    case legacyPreRule = "legacy_pre_rule"
}

/// Arbitrary JSON payload (schema keeps payload an open object).
public enum JSONValue: Codable, Sendable, Equatable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    public init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
        } else if let b = try? c.decode(Bool.self) {
            self = .bool(b)
        } else if let n = try? c.decode(Double.self) {
            self = .number(n)
        } else if let s = try? c.decode(String.self) {
            self = .string(s)
        } else if let a = try? c.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? c.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(in: c, debugDescription: "unsupported JSON value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .null: try c.encodeNil()
        case .bool(let b): try c.encode(b)
        case .number(let n): try c.encode(n)
        case .string(let s): try c.encode(s)
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        }
    }

    public subscript(key: String) -> JSONValue? {
        if case .object(let o) = self { return o[key] }
        return nil
    }

    public var stringValue: String? {
        if case .string(let s) = self { return s }
        return nil
    }

    public var boolValue: Bool? {
        if case .bool(let b) = self { return b }
        return nil
    }

    public var numberValue: Double? {
        if case .number(let n) = self { return n }
        return nil
    }
}
